gto/git_utils.py

```diff
@@ -67,3 +68,26 @@ def cloned_git_repo(repo: str):
 def git_clone(repo: str, dir: str) -> None:
     logging.debug("clone %s in directory %s", repo, dir)
     Repo.clone_from(url=repo, to_path=dir)
+
```

Comment:
Something to consider: it could be that we have some Git tags locally, but don't have them in remote. Then if you run `--auto-push` in locally cloned repo, you'll decide what to do based on the local repo, but you will try to apply it to the remote one. It could be an error in some cases (you try to remove a remote Git tag that doesn't exist in remote). It could be a incorrect decision (you deprecate a version that doesn't exist in the remote). Looks too complicated to resolve at Friday evening 🍻

Reply:
"From great power comes great responsibility" I'd say :D
In the end the auto-push is something you can choose to use or not at every call, as you can choose or not to git push a tag. But if this were not (very) useful, why would you echo a suggestion to push the created tags?

In practice I find hard to imagine to work with a local repo out of sync with the remote. This is the CI/CD practice, isn't it?
In fact I would suggest to be able to set auto-push always true in a config file...

Reply:
> In practice I find hard to imagine to work with a local repo out of sync with the remote.

This could happen by chance. But ok, I guess we'll need to accommodate for this in the future. This may happen as part of implementing `gto status` command, so I'll write it [in that ticket](https://github.com/iterative/gto/issues/115#issue-1201663040).

> In fact I would suggest to be able to set auto-push always true in a config file...

Good idea!

Reply:
Example of the currently misleading behavior. Consider you have a git tag `bug@v0.0.1` in remote, but don't have it locally. Then:

```
$ gto register bug --auto-push
Created git tag 'bug@v0.0.1' that registers version
Running `git push origin bug@v0.0.1`
WARNING:git.remote:Error lines received while fetching: error: failed to push some refs to 'https://github.com/aguschin/example-gto'
Successfully pushed git tag bug@v0.0.1 on remote.
```

I think we should fail at push, printing something like: `The Git tag bug@v0.0.1 already exists on remote.` WDYT?

Reply:
I agree that it would be more user friendly, therefore desirable. But on the other hand, the same exact error would happen if the user does the git push manually, right? So, as the user can choose if to do git push or not, can also choose whether to add the option `--auto-push` or not.

In conclusion, I don't see much harm in it, but one could make error messages easier to interpret, I agree.

Reply:
Shall we create a new issue for this?

Reply:
Ok, think I got your point. As I see the difference
1. We're communicating it incorrectly. We say `Successfully pushed git tag...`. E.g. the user may assume the CI is triggered in the remote. But the tag was not pushed, and CI was not triggered. So we're confusing user with this message.
2. Often, a user registers a version or assigns a stage to trigger CI. In this case the goal wasn't achieved, so we should exit with code 1. It could be that user don't want any CI and just want the remote repo to be in this state (so as the Git tag was pushed, we don't need to do anything and it is fine). In this case it may be ok, but I think triggering the CI is more frequent goal here. We may add some tag (`--no-fail-if-tags-exist-in-remote` e.g.), but I think it's overkill now.

Reply:
ok, then is it ok if I apply the following change: if `git push` fails, then echo an appropriate message and exit with code 1?

Reply:
appropriate message being something like "Execution of `git push origin bug@v0.0.1` failed. Make sure your local repository is in sync with the remote."

ok?

Reply:
Yes, totally. Thanks 🙏🏻

Reply:
This should be it: https://github.com/iterative/gto/pull/281/commits/0f39722686ab42f29a3d60de37f904089e013975